frontend/src/routes/password.change.tsx

```diff
@@ -0,0 +1,283 @@
+import IconLockSolid from "@vector-im/compound-design-tokens/icons/lock-solid.svg?react";
+import { Alert, Form, Separator } from "@vector-im/compound-web";
+import { FormEvent, useRef } from "react";
+import { useTranslation } from "react-i18next";
+import { useMutation, useQuery } from "urql";
+
+import BlockList from "../components/BlockList";
+import { ButtonLink } from "../components/ButtonLink";
+import Layout from "../components/Layout";
+import LoadingSpinner from "../components/LoadingSpinner";
+import PageHeading from "../components/PageHeading";
+import { graphql } from "../gql";
+import { SetPasswordStatus } from "../gql/graphql";
+
+const CURRENT_VIEWER_QUERY = graphql(/* GraphQL */ `
+  query CurrentViewerQuery {
+    viewer {
+      __typename
+      ... on Node {
+        id
+      }
+    }
+  }
+`);
+
+const CHANGE_PASSWORD_MUTATION = graphql(/* GraphQL */ `
+  mutation ChangePassword(
+    $userId: ID!
+    $oldPassword: String!
+    $newPassword: String!
+  ) {
+    setPassword(
+      input: {
+        userId: $userId
+        currentPassword: $oldPassword
+        newPassword: $newPassword
+      }
+    ) {
+      status
+    }
+  }
+`);
+
+export const Route = createFileRoute("/password/change")({
+  async loader({ context, abortController: { signal } }) {
+    const viewer = await context.client.query(
+      CURRENT_VIEWER_QUERY,
+      {},
+      { fetchOptions: { signal } },
+    );
+    if (viewer.error) throw viewer.error;
+    if (viewer.data?.viewer.__typename !== "User") throw notFound();
+  },
+
+  component: ChangePassword,
+});
+
+function ChangePassword(): React.ReactNode {
+  const { t } = useTranslation();
+  const [viewer] = useQuery({ query: CURRENT_VIEWER_QUERY });
+  const router = useRouter();
+  if (viewer.error) throw viewer.error;
+  if (viewer.data?.viewer.__typename !== "User") throw notFound();
+  const userId = viewer.data.viewer.id;
+
+  const currentPasswordRef = useRef<HTMLInputElement>(null);
+  const newPasswordRef = useRef<HTMLInputElement>(null);
+  const newPasswordAgainRef = useRef<HTMLInputElement>(null);
+
+  const [result, changePassword] = useMutation(CHANGE_PASSWORD_MUTATION);
+
+  const onSubmit = async (event: FormEvent<HTMLFormElement>): Promise<void> => {
+    event.preventDefault();
+
+    const formData = new FormData(event.currentTarget);
+
+    const oldPassword = formData.get("current_password") as string;
+    const newPassword = formData.get("new_password") as string;
+    const newPasswordAgain = formData.get("new_password_again") as string;
+
+    if (newPassword !== newPasswordAgain) {
+      throw new Error("passwords mismatch; this should be checked by the form");
+    }
+
+    const response = await changePassword({ userId, oldPassword, newPassword });
+
+    if (response.data?.setPassword.status === SetPasswordStatus.Allowed) {
+      router.navigate({ to: "/password/change_success" });
```

Comment:
I'm nitpicking at this point, but I personally don't like underscores in paths, and we used dashes elsewhere. Can you make this either `/password/change/success` or `/password/change-success`?

Reply:
heh I am the opposite, I don't like the dashes in paths.

But anyway, I tried `/` but I must have been doing something wrong — it just renders the change password form instead of the success page so I gave up on that. Will change to dashes

Reply:
that's because of how the router nests components

`password.change.success.tsx` will basically render `Password > Change > Success`, nesting components using the `<Outlet />` thingy.

You'd need to move the `password.change.tsx` page to `password.change.index.tsx`

see: https://tanstack.com/router/latest/docs/framework/react/guide/route-trees#flat-routes

Reply:
aaah. That makes more sense, but still grrr :-). I will try that

Reply:
yeah that appears to work